src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp

```diff
@@ -185,192 +185,217 @@ Pass Orbit::getPassWithRevolutionNumber(const Integer& aRevolutionNumber, const
+
+    Pass const* closestPasSPtr = nullptr;
 
-        const auto lowerBoundMapIt = this->passMap_.lower_bound(aRevolutionNumber);
+    const auto lowerBoundMapIt = this->passMap_.lower_bound(aRevolutionNumber);
 
-        if (lowerBoundMapIt != this->passMap_.end())
+    if (lowerBoundMapIt != this->passMap_.end())
+    {
+        if (lowerBoundMapIt == this->passMap_.begin())
+        {
+            closestPasSPtr = &(lowerBoundMapIt->second);
+        }
+        else
         {
-            if (lowerBoundMapIt == this->passMap_.begin())
+            const auto closestPassMapIt = std::prev(lowerBoundMapIt);
+
+            if ((aRevolutionNumber - closestPassMapIt->first) < (lowerBoundMapIt->first - aRevolutionNumber))
             {
-                closestPasSPtr = &(lowerBoundMapIt->second);
+                closestPasSPtr = &(closestPassMapIt->second);
             }
             else
             {
-                const auto closestPassMapIt = std::prev(lowerBoundMapIt);
-
-                if ((aRevolutionNumber - closestPassMapIt->first) < (lowerBoundMapIt->first - aRevolutionNumber))
-                {
-                    closestPasSPtr = &(closestPassMapIt->second);
-                }
-                else
-                {
-                    closestPasSPtr = &(lowerBoundMapIt->second);
-                }
+                closestPasSPtr = &(lowerBoundMapIt->second);
             }
         }
-        else if (this->passMap_.size() > 0)
+    }
+    else if (this->passMap_.size() > 0)
+    {
+        closestPasSPtr = &(this->passMap_.begin()->second);
+    }
+
+    Pass currentPass = Pass::Undefined();
+
+    if (closestPasSPtr != nullptr)  // Pass with closest revolution number found
+    {
+        currentPass = *closestPasSPtr;
+    }
+
+    Integer currentRevolutionNumber =
+        currentPass.isDefined() ? currentPass.getRevolutionNumber() : this->modelPtr_->getRevolutionNumberAtEpoch();
+
+    const bool isForwardPropagated = currentRevolutionNumber <= aRevolutionNumber;
+    const Integer propagationSign = isForwardPropagated ? 1 : -1;
+
+    const Instant epoch = this->modelPtr_->getEpoch();
+
+    const auto getZ = [this, &epoch](const double& aDurationInSeconds) -> Real
+    {
+        return this->modelPtr_->calculateStateAt(epoch + Duration::Seconds(aDurationInSeconds))
+            .getPosition()
+            .accessCoordinates()
+            .z();
+    };
+
+    const auto getZDot = [this, &epoch](const double& aDurationInSeconds) -> Real
+    {
+        return this->modelPtr_->calculateStateAt(epoch + Duration::Seconds(aDurationInSeconds))
+            .getVelocity()
+            .accessCoordinates()
+            .z();
+    };
+
+    while ((!currentPass.isDefined()) || (currentPass.getRevolutionNumber() != aRevolutionNumber))
+    {
+        Instant previousInstant = Instant::Undefined();
+
+        Duration stepDuration = aStepDuration;
+        if (currentPass.isDefined() && currentPass.isComplete())
```

Comment:
Small Q:

Is the `currentPass.isComplete()` still necessary? If we don't save incomplete passes, then can't we assume that if a pass is defined that it is also complete?

Reply:
Good catch!